wolfssl/src/chacha20_poly1305.rs

```diff
@@ -0,0 +1,108 @@
+use bytes::BytesMut;
+use wolfssl_sys::{wc_ChaCha20Poly1305_Decrypt, wc_ChaCha20Poly1305_Encrypt};
+
+/// Struct for encrypt/decrypt using Chacha20 cipher and authentication
+/// using Poly1305
+pub struct Chacha20Poly1305Aead {
+    key: [u8; Self::KEY_SIZE],
+}
+
+impl Chacha20Poly1305Aead {
+    /// Size of Initialisation vector
+    pub const IV_SIZE: usize = wolfssl_sys::CHACHA20_POLY1305_AEAD_IV_SIZE as usize;
+
+    /// Size of key
+    pub const KEY_SIZE: usize = wolfssl_sys::CHACHA20_POLY1305_AEAD_KEYSIZE as usize;
+
+    /// Size of auth tag
+    pub const AUTHTAG_SIZE: usize = wolfssl_sys::CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE as usize;
+
+    /// Creates new `Chacha20Poly1305Aead` instance
+    pub fn new(key: [u8; Self::KEY_SIZE]) -> Self {
+        Self { key }
+    }
+
+    /// This function encrypts an input message `plain_text`, using the ChaCha20 stream cipher,
+    /// It also performs Poly-1305 authentication (on the cipher text),
+    /// and stores the generated authentication tag in the output buffer
+    pub fn encrypt(
+        &self,
+        iv: [u8; Self::IV_SIZE],
+        plain_text: &[u8],
+    ) -> Result<(BytesMut, [u8; Self::AUTHTAG_SIZE]), String> {
+        let mut cipher_text = BytesMut::zeroed(plain_text.len());
```

Comment:
You could use `BytesMut::with_capacity()` here and then [`cipher_text.spare_capacity_mut`](https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.spare_capacity_mut) here and avoid bothering to initialize the input buffer just to be overwritten in the FFI call.

`Session<IOCB>::try_read` has an example.

Reply:
`BytesMut::spare_capacity_mut` needed unsafe code to set_len after populating with data, so avoided it. But who am i kidding.
It's already unsafe to call C apis and it is in data path. Updated it to use `spare_capacity_mut`